### PR TITLE
Add initial CI workflow

### PR DESCRIPTION
**Pull request by eproxus**

### CODE REVIEW

**Comment by eproxus** [comment]
@c-mauderer Seems [my Linux check](https://github.com/grisp/grisp2-rtems-toolchain/blob/master/Makefile#L106) doesn't work. Can you try it on your machine?

**Comment by c-mauderer** [comment]
I think your indentation in the makefile is wrong. You must not indent the ifneq. Otherwise make will try to execute it as a command. Beneath that I think you missed the endif. Try

```
.PHONY: barebox
#H Build the bootloader
barebox:
ifneq ($(UNAME),Linux)
	$(error "Barebox can only be built on Linux")
else
	cd $(SRC_BAREBOX) && rm -f .config
	cd $(SRC_BAREBOX) && ln -s $(MAKEFILE_DIR)/barebox/config .config
	cd $(SRC_BAREBOX) && make ARCH=arm CROSS_COMPILE=arm-rtems5- -j$(NUMCORE)
endif
```

That worked on my system.

**Comment by eproxus** [comment]
🤦 Of course, it's fixed on `master` now

**Comment by eproxus** [comment]
@c-mauderer Now the compilation fails on Linux with an error when building Barebox. Is this something you've seen before?

```
11175
  HOSTCC  scripts/bareboxcrc32
11176
  HOSTLD  scripts/dtc/dtc
11177
  HOSTCC  scripts/kernel-install
11178
  HOSTCC  scripts/bareboximd
11179
In file included from /usr/include/fcntl.h:290:0,
11180
                 from scripts/bareboximd.c:24:
11181
In function ‘open’,
11182
    inlined from ‘write_file’ at scripts/bareboximd.c:62:5,
11183
    inlined from ‘imd_write_crc32’ at scripts/../common/imd.c:380:4,
11184
    inlined from ‘imd_command’ at scripts/../common/imd.c:487:3:
11185
/usr/include/x86_64-linux-gnu/bits/fcntl2.h:50:4: error: call to ‘__open_missing_mode’ declared with attribute error: open with O_CREAT or O_TMPFILE in second argument needs 3 arguments
11186
    __open_missing_mode ();
11187
    ^~~~~~~~~~~~~~~~~~~~~~
11188
  HOSTCC  scripts/kallsyms
11189
make[2]: *** [scripts/bareboximd] Error 1
11190
make[2]: *** Waiting for unfinished jobs....
11191
scripts/Makefile.host:91: recipe for target 'scripts/bareboximd' failed
11192
make[1]: *** [scripts] Error 2
11193
Makefile:436: recipe for target 'scripts' failed
11194
make: *** [barebox] Error 2
11195
make[1]: Leaving directory '/home/runner/work/grisp2-rtems-toolchain/grisp2-rtems-toolchain/external/barebox'
11196
Makefile:109: recipe for target 'barebox' failed
11197
##[error]Process completed with exit code 2.
```

**Comment by c-mauderer** [comment]
No, I haven't seen that kind of error on my machine. It seems a bit odd. the `bareboximd` is a command for the development host. I don't really know what it does. The help tells me that it `Extract metadata from a barebox image`. I don't think that we really need it.

The really odd part about your error is that it seems to be missing some system function. It's something in `open`.

A ldd on the tool gives me the following output:
```
 > ldd scripts/bareboximd
        linux-vdso.so.1 (0x00007ffd43560000)
        libc.so.6 => /lib64/libc.so.6 (0x00007ff8640b2000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff86446d000)
```
So not a lot of dependencies either.

Maybe just try to disable it. It should be the following change:
```
diff --git a/barebox/config b/barebox/config
index c7a5990..0b675d2 100644
--- a/barebox/config
+++ b/barebox/config
@@ -312,7 +312,7 @@ CONFIG_CMD_DRVINFO=y
 CONFIG_CMD_HELP=y
 CONFIG_LONGHELP=y
 CONFIG_CMD_IOMEM=y
-CONFIG_CMD_IMD=y
+# CONFIG_CMD_IMD is not set
 CONFIG_CMD_MEMINFO=y
 CONFIG_CMD_ARM_MMUINFO=y
 CONFIG_CMD_REGULATOR=y
```

**Comment by eproxus** [comment]
Still can't make this build on Linux. Error at the bottom here: https://github.com/grisp/grisp2-rtems-toolchain/commit/474c1e90a5439bfbff3455b1a75119fc207039f9/checks/662737012/logs

```
2020-05-11T10:32:37.1151251Z   HOSTCC  scripts/kallsyms
2020-05-11T10:32:37.1550789Z In file included from /usr/include/fcntl.h:290:0,
2020-05-11T10:32:37.1551236Z                  from scripts/bareboximd.c:24:
2020-05-11T10:32:37.1555507Z In function â€˜openâ€™,
2020-05-11T10:32:37.1555867Z     inlined from â€˜write_fileâ€™ at scripts/bareboximd.c:62:5,
2020-05-11T10:32:37.1556361Z     inlined from â€˜imd_write_crc32â€™ at scripts/../common/imd.c:380:4,
2020-05-11T10:32:37.1559618Z     inlined from â€˜imd_commandâ€™ at scripts/../common/imd.c:487:3:
2020-05-11T10:32:37.1560385Z /usr/include/x86_64-linux-gnu/bits/fcntl2.h:50:4: error: call to â€˜__open_missing_modeâ€™ declared with attribute error: open with O_CREAT or O_TMPFILE in second argument needs 3 arguments
2020-05-11T10:32:37.1560699Z     __open_missing_mode ();
2020-05-11T10:32:37.1560867Z     ^~~~~~~~~~~~~~~~~~~~~~
2020-05-11T10:32:37.2030369Z make[2]: *** [scripts/bareboximd] Error 1
```

Not sure what could be wrong here. Any ideas @c-mauderer?

**Comment by c-mauderer** [comment]
That's still the bareboximd.c file with the same error message. It seems that disabling that tool didn't work so I had a look at the problem.

I tried building barebox on a Ubuntu 18.04 VM to be as close as possible to the CI system used by github. They have a really picky C-Library version. The problematic line in bareboximd.c is:

```
fd = open(filename, O_WRONLY | O_TRUNC | O_CREAT);
```
The error message tells us that an argument is missing if O_CREAT is uesd. And really: If you read the man page of open:

>       O_CREAT       If  the  file exists, this flag has no effect [...]
>                     and the access permission bits (see <sys/stat.h>) of the file  mode shall  be
>                     set to the value of the argument following the oflag argument taken as type
>                     mode_t modified as follows [...]

So if the line is replaced by
```
fd = open(filename, O_WRONLY | O_TRUNC | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
```
everything works. I pushed a patch to v2020.03.0-GRiSP2 branch in the barebox repo. But that sounds like something that should be done as an upstream fix. I'll search whether they have a bug tracker or similar.

**Comment by c-mauderer** [comment]
PS: Please use https://github.com/grisp/barebox/commit/e2d1530f2c5f4bf6fcf735480a511adc495f4630 for the barebox submodule.

**Comment by c-mauderer** [comment]
> But that sounds like something that should be done as an upstream fix. I'll search whether they have a bug tracker or similar.

Just for your information: I posted the patch on the barebox mailinglist and received the information that there is already a identical patch on the master. Seems that I missed that.

**Comment by eproxus** [comment]
@c-mauderer Thanks! Is it safe to switch to master or do we have patches in our fork we need?

**Comment by c-mauderer** [comment]
We need that patch: https://github.com/grisp/barebox/commit/51804e4e3bc8e2bfb34a9784b955c79eb15f6ab1
That is currently a bit hacked in and therefore I didn't push it upstream. I'm not sure yet whether the FDT is really complete enough. Beneath that some of the locations that I touched are quite phytec centered and maybe would need an extra category for this project / company instead.

**Comment by c-mauderer** [comment]
Oh, and I don't think that our target for the Bootloader should be to always use the latest version. Instead a stable and tested version would be a better target. Therefore I forked off at a version tag instead of some random master commit. I'm not sure how well tested the master is in the barebox project.

**Comment by eproxus** [comment]
If you update the submodule with the fix, I can try CI again.

**Comment by c-mauderer** [comment]
I thought you would just update the submodule to https://github.com/grisp/barebox/commit/e2d1530f2c5f4bf6fcf735480a511adc495f4630 in your CI branch. That is a identical commit to the fix that has been done on upstream/master. Only with a different message because I created it without noting that there already exists one.

File: barebox/config [hunk 1]
```diff
@@ -312,7 +312,7 @@ CONFIG_CMD_DRVINFO=y
 CONFIG_CMD_HELP=y
 CONFIG_LONGHELP=y
 CONFIG_CMD_IOMEM=y
-CONFIG_CMD_IMD=y
+# CONFIG_CMD_IMD is not set
```
Comment (c-mauderer) [inline barebox/config:315]: I think that change shouldn't be necessary any more. Building the command should be fixed with the patch on our barebox branch (that you pulled in already).

Comment (eproxus) [reply barebox/config:315]: Ah, good point! I'll remove it

**Comment by eproxus** [comment]
@c-mauderer Ready for review! The two commits are independent, so we can rebase and merge when it is approved.

**Comment by eproxus** [comment]
@c-mauderer What is needed in the README apart from the dependency information already added?

**Comment by c-mauderer** [comment]
Sorry: Missunderstanding: The only differences between this PR and the master are:
- Your CI receipe
- Changes to README
- Update of the barebox repo

The other changes from earlier (like the fix for disabling barebox build on MacOS) have already been merged earlier.

**Comment by c-mauderer** [comment]
Means: The changes are all non-critical ones. Please merge them.

**Pull request merged**